### PR TITLE
Proposal to change linknames for dependencies on AIX.

### LINKED ISSUE #6368: Linking issues on AIX with OpenSSL 1.0.2 and 1.1.0 in the same location
It's probably not meant to be that way, but I wasn't aware that AIX is lacking any concept of a SONAME support and I have a decision to have OpenSSL 1.0.2 and 1.1.0 installed side-by-side, which isn't a problem on any other platform I have to look at.

The issue at hand now is, that with a linker instructions for the aix64-cc target the libssl.so.1.1 gets a dependency to libcrypto.so instead of libcrypto.so.1.1. Depending on the way I configure the side-by-side installation this either leads to no problem - if libcrypto.so points to libcrypto.so.1.1 - or to unresolved symbols - if libcrypto.so points to libcrypto.so.1.0.0 - or an error because libcrypto.so could not be found.

I partly resolved this already. For the applications that get ported to OpenSSL 1.1 I instructed the developers to use absolute paths when linking to OpenSSL 1.1, i.e. ```-brtl $(OPENSSL_LIB)/libcrypto.so.1.1 $(OPENSSL_LIB)/libssl.so.1.1``` instead of using ```-brtl -L $(OPENSSL_LIB) -lcrypto -lssl```. This is doing the right thing, as ld is removing the path component and just logs the dependency to libcrypto.so.1.1 and libssl.so.1.1 and proved to work out for executables that only need the crypto-library.

The OpenSSL build environment doesn't do such a stunt of course which results in above mentioned dependency. Therefore applications that need the SSL-library are facing unresolved symbols, because the old libcrypto.so from 1.0.2 gets loaded.

Is there a way to instruct the OpenSSL build to do a linking as I proposed to our application developers? I don't see a way to get this through into link_shlib.aix, but then I don't fully grok this mechanic anyway. When libssl.so is created the ```-lcrypto``` would need to be replaced by the absolute path to libcrypto.so.1.1...

May I also propose a name change for the libraries on AIX to be like on Windows - libcrypto-1.1.so and libssl-1.1.so - or at least have symlinks with that name?

kind regards
-- Matthias Kraft

### PR DESCRIPTION
**Pull request by makr**
Directly link against the full library names on AIX in order to enable
having multiple library versions in the same directory.

Signed-off-by: Matthias Kraft <Matthias.Kraft@softwareag.com>

Fixes #6368

### CODE REVIEW

File: Makefile.shared [hunk 1]
```diff
@@ -430,24 +430,36 @@ link_app.hpux:
...
 	$(LINK_APP)
 
+# AIX has no concept of SONAME, hence using -lname would always link
+# against libname.so without any notion of a version. The ld command
+# accepts full names in addition to -l, however. With -bnoipath the path
+# component is dropped, so we get a similar behavior as on other
+# platforms.
+# The sed command is limited in its RE support, so instead of addressing
+# words with "-l\(crypto\|ssl\)" we have to address single characters
+# with "-l\([cs][rs][yl]p*t*o*\)" for the same effect.
```
Comment (levitte) [inline Makefile.shared]: Oh wow...

File: Configurations/unix-Makefile.tmpl [hunk 2]
```diff
@@ -912,12 +912,14 @@ EOF
...
 # {libname}$shlibextsimple for the Unix platforms.
+# The AIX targets in Makefile.shared need SHLIB_MAJOR and SHLIB_MINOR.
 $target: $lib$libext $deps $ordinalsfile
 	\$(MAKE) -f \$(SRCDIR)/Makefile.shared -e \\
 		PLATFORM=\$(PLATFORM) \\
 		PERL="\$(PERL)" SRCDIR='\$(SRCDIR)' DSTDIR="$libd" \\
 		INSTALLTOP='\$(INSTALLTOP)' LIBDIR='\$(LIBDIR)' \\
 		LIBDEPS='\$(PLIB_LDFLAGS) '"$linklibs"' \$(EX_LIBS)' \\
+		SHLIB_MAJOR='\$(SHLIB_MAJOR)' SHLIB_MINOR='\$(SHLIB_MINOR)' \\
 		LIBNAME=$libname SHLIBVERSION=\$(SHLIB_MAJOR).\$(SHLIB_MINOR) \\
```
Comment (levitte) [inline Configurations/unix-Makefile.tmpl]: Did you notice `SHLIBVERSION` here?  I don't see a need to add another two variables...

File: Configurations/unix-Makefile.tmpl [hunk 3]
```diff
@@ -954,6 +956,7 @@ $target: $objs $deps
 		PLATFORM=\$(PLATFORM) \\
 		PERL="\$(PERL)" SRCDIR='\$(SRCDIR)' DSTDIR="$dsod" \\
 		LIBDEPS='\$(PLIB_LDFLAGS) '"$shlibdeps"' \$(EX_LIBS)' \\
+		SHLIB_MAJOR='\$(SHLIB_MAJOR)' SHLIB_MINOR='\$(SHLIB_MINOR)' \\
```
Comment (levitte) [inline Configurations/unix-Makefile.tmpl]: It would be simpler to set `SHLIBVERSION`, just as bove

File: Configurations/unix-Makefile.tmpl [hunk 4]
```diff
@@ -991,6 +994,7 @@ $bin$exeext: $objs $deps
 		PERL="\$(PERL)" SRCDIR=\$(SRCDIR) \\
 		APPNAME=$bin$exeext OBJECTS="$objs" \\
 		LIBDEPS='\$(PLIB_LDFLAGS) '"$linklibs"' \$(EX_LIBS)' \\
+		SHLIB_MAJOR='\$(SHLIB_MAJOR)' SHLIB_MINOR='\$(SHLIB_MINOR)' \\
```
Comment (levitte) [inline Configurations/unix-Makefile.tmpl]: `SHLIBVERSION` here as well?

File: Makefile.shared [hunk 5]
```diff
@@ -430,24 +430,36 @@ link_app.hpux:
...
+#
 link_dso.aix:
 	@OBJECT_MODE=`expr "x$(SHARED_LDFLAGS)" : 'x\-[a-z]*\(64\)'` || :; \
 	OBJECT_MODE=$${OBJECT_MODE:-32}; export OBJECT_MODE; \
 	ALLSYMSFLAGS=''; \
 	NOALLSYMSFLAGS=''; \
-	SHAREDFLAGS='$(CFLAGS) $(SHARED_LDFLAGS) -Wl,-bexpall,-bnolibpath,-bM:SRE'; \
+	SHAREDFLAGS='$(CFLAGS) $(SHARED_LDFLAGS) -Wl,-bexpall,-bnolibpath,-bnoipath,-bM:SRE'; \
 	rm -f $(SHLIBNAME_FULL) 2>&1 > /dev/null ; \
+	LIBDEPS=`echo $(LIBDEPS) | sed -e 's@-L\([^ ]*\) -l\([cs][rs][yl]p*t*o*\)@\1/lib\2.so.$(SHLIB_MAJOR).$(SHLIB_MINOR)@g'` || :;\
```
Comment (levitte) [inline Makefile.shared]: use `$(SHLIBVERSION)` instead of `$(SHLIB_MAJOR).$(SHLIB_MINOR)`

**Comment by dot-asm** [comment]
Question is not only how openssl executable is linked during build, but even how applications are supposed to link against installed libraries. Customarily people expect -L /some/where -lcrypto, right? And what is suggested here? /some/where/libcrypto.so.1.1, i.e. by explicit *full* path. Would people actually appreciate it? I wonder if -L /some/where -lcrypto-1.1 would be more appropriate? Provided that we accept the premise that version has to be in the shared library's name that is. Because otherwise one can also wonder if it would be reasonable to *demand* that 1.0.x and 1.1.x are installed in different locations *and* binaries would have to look for shared libraries at these different locations. Note that I'm not arguing in favour of the latter, I only wonder what would be most reasonable thing to do. Because it might be something totally different...

It also makes me wonder how it worked so far. We install both libcrypto.a, libcrypto.so and libcrypto.so.ver into target location. But AIX linker's search algorithm for -lname seems to be "look for libname.a, if not found[!], look for libname.so". In which case -L /pre/fix/lib -lcrypto will always link with ... static library. Ah! Unless you pass -Wl,-bsvr4, in which case it looks for .so first. [And that's when equivalent of SONAME becomes handy.] But then question is what are AIX application developers/maintainers accustomed to. In other words it seems that in order to choose what to do, we have to assume something quite specific about AIX application developers/maintainers... The situation is intensified by the way AIX handles own shared libraries. They put multiple shared libraries into .a file, even of different bitness. So that ldd says for example /usr/lib/libc.a(shr.o) or /usr/lib/libc.a(shr_64.o), depending on if it's 32- or 64-bit application. It sounds that most sensible thing to do would be to

1. install static libraries in /pre/fix/lib/static/lib{crypto|ssl}.a
2. create /pre/fix/lib/lib{crypto|ssl}.a with libcrypto.so.ver inside.

And tell those who are accustomed to and expect static link to modify -L flag as -L/some/where/static. Trouble is that it would take custom AIX-specific install procedure, which contradicts the desire to unify all Unices...

**Comment by dot-asm** [comment]
It should be noted that AIX linker is very lax about what can be used as input. In fact it's apparently possible to use it to relink applications. Imagine you have apps/openssl linked with libssl.so and libcrypto.so. Now you `ar r /tmp/libopenssl.a libcrypto.so.1.1 libssl.so.1.1`. And now you `ld -o /tmp/openssl apps/openssl -L /tmp -lopenssl -lc -lpthread`. It works and ldd says all the right things. Well, binary itself can't be started, because /tmp/libopenssl.a(libssl.so.1.1) has dependency on libcrypto.so. But it should be possible to resolve. Bottom line is that ld apparently can be used as "install" tool. Or even post-install...

**Comment by dot-asm** [comment]
On related side note AIX linker seems to have -bsymbolic. And it would be appropriate to use it...

**Comment by dot-asm** [comment]
On another related side note AIX linker seems to have a notion of export list. As of now we -bexpall, but apparently it doesn't have to be the case.

**Comment by dot-asm** [comment]
> /pre/fix/lib/static/lib{crypto|ssl}.a

One can actually wonder if it would be appropriate to do so on all Unices (and why not even Windows ;-). Naturally in shared build. I mean if configured with no-shared then it .a should still be installed in lib. Thing is that if you have .so's and want to link with .a instead, you'd have to specify them by full path, which is not exactly practical... No?

**Comment by levitte** [comment]
Hmmmm? On most unixen, static linking is all about using the right flag (`-static`, `-Wl,-static`, `-Bstatic`,...), is it not? I don't see why you'd need to give full paths in the general case...

**Comment by dot-asm** [comment]
> On most unixen, static linking is all about using the right flag (-static ...

Trouble is that it's all-or-nothing flag. I mean -static says link *all* libraries static, while question is rather what if libcrypto.a is the only one I want to link statically.

File: Makefile.shared (hunk 5 above)
Comment (makr) [reply Makefile.shared]: Damn heat! I don't know what I was thinking. Especially as I started experimenting using SHLIBVERSION, _because_ it was already available for the shlib build. I'll change that when back at work next week ... Sorry.

**Comment by dot-asm** [comment]
> And what is suggested here? /some/where/libcrypto.so.1.1, i.e. by explicit full path.

I view this as least appropriate option. Which means that I'd argue against this suggestion. BTW, compiler issues warning which reads "contains an **incorrect** file suffix" if you attempt to link with .so.ver, i.e. I kind of have compiler on my side :-)

**Comment by dot-asm** [comment]
>> On most unixen, static linking is all about using the right flag (-static ...

> Trouble is that it's all-or-nothing flag. I mean -static says link all libraries static, while question is rather what if libcrypto.a is the only one I want to link statically.

This turned to be untrue. In sense that you can alternate between static and dynamic libraries for example as `-Wl,-Bstatic,-lcrypto,-Bdynamic`. Well, question would still be if it's universally available option, but suggestion to place static libraries into dedicated directory on all systems should be disregarded. Sorry about possible confusion.

**Comment by dot-asm** [comment]
Since we effectively decided to go for libcrypto.a(libcrypto[64].so.1.1) approach, this is effectively obsoleted by #6487. So I'm closing this. However, I'd appreciate if you, Mattias, could file separate request for the dso_dlfcn.c modification that accommodates new convention. Targeting master as initial step.

**Comment by makr** [comment]
I will need a while, currently very busy and then vacation. So it will probably not be before end of July. However, there is commit makr/openssl@495cbb4279578c9d5a7b5268339017df85d5a14b which does the `dladdr()` implementation already. Outstanding is a working shlibloadtest.

**Pull request closed without merging**